### PR TITLE
#160577582 Implement follow/unfollow and viewing of other users profiles

### PR DESCRIPTION
**Pull request by C3real-kill3r**
**What does this PR do?**

implement follow/unfollow feature and check other users profile

**Description of Task to be completed?**

- enables a users to follow and unfollow other users
- enable users to view the list of their followed users and users following them

**How should this be manually tested?**

* Open this PR's review app.
* Click on an article to open it.(not authored by you)
* click on the follow button to either(follow when green/unfollow when red)
* go to the top-right icon with your profile picture and click on it
* click profiles and navigate through the followers and following tab to view the respective users

**Any background context you want to provide?**

N/A

**What are the relevant pivotal tracker stories?**

[#160577582](https://www.pivotaltracker.com/story/show/160577582)

**Screenshots (if appropriate)**

* list of a follower
screenshot 2018-11-23 at 07 31 33

* an article of a user you're not following
screenshot 2018-11-23 at 07 32 02

*an article of  a user after following them
screenshot 2018-11-23 at 07 32 14

*a list of users you are following
screenshot 2018-11-23 at 07 33 07

**Questions:**

N/A

### CODE REVIEW

**Review by YAHYA-H — APPROVED** [review]
Impressive!!!! 
Good job 😄 👍

File: src/components/ProfileView/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import profileImage from '../../assets/images/profile.jpg';
+import './ProfileView.scss';
+import ConnnectedFollowUnfollow from "../../containers/FollowUnfollow";
+import { getCurrentUser } from "../../utils/auth";
+
+const ProfileView = ({ profile }) => (
+  <>
```
Comment (paulrimiru) [inline src/components/ProfileView/index.js:8]: Add snapshot test for this component

File: src/containers/ArticleListing/state/reducer.js [hunk 2]
```diff
@@ -19,33 +19,33 @@ export const initialState = {
...
+    };
+  }
+  case FETCH_ARTICLES_FAILURE:
+    return {
+      ...state,
+      errors: action.payload.errors,
+      isFetching: false,
+    };
+  default:
+    return state;
```
Comment (paulrimiru) [inline src/containers/ArticleListing/state/reducer.js:48]: Add test for the default, check if the state is empty first in order to trigger this line

File: src/containers/FollowButton/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    fetch(username);
+  };
+
+  handleFollow = () => {
+    const { follow, username } = this.props;
+    follow(username);
+  };
+
+  handleUnfollow = () => {
+    const { unfollow, username } = this.props;
```
Comment (paulrimiru) [inline src/containers/FollowButton/index.js:18]: Add test for this function check if the prop is called with the correct arguments

File: src/containers/FollowUnfollow/index.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  };
+
+  handleFollow = () => {
+    const { follow, username } = this.props;
+    follow(username);
+  };
+
+  handleUnfollow = () => {
+    const { unfollow, username } = this.props;
+    unfollow(username);
```
Comment (paulrimiru) [inline src/containers/FollowUnfollow/index.js:19]: Same case here

File: src/containers/FollowUnfollow/index.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  };
+
+  render() {
+    const { username } = this.props;
+    let isFollowing = false;
+    if (this.props && this.props.users) {
+      // eslint-disable-next-line
+      for (const user of this.props.users.following.users) {
```
Comment (paulrimiru) [inline src/containers/FollowUnfollow/index.js]: Add test for this functionality ensure that you pass `users` prop in order to trigger these lines

File: src/containers/FollowUnfollow/state/actions.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  error,
+});
+
+export const fetchAction = username => (dispatch) => {
+  dispatch(pageLoadingAction());
+  return api.get(`profiles/following/${username}`)
+    .then((res) => {
+      const following = res.data.following_status;
+      if (following) {
+        dispatch(checkFollowers(res.data));
```
Comment (paulrimiru) [inline src/containers/FollowUnfollow/state/actions.js:50]: Add test for if following is true

File: src/containers/FollowUnfollow/state/actions.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    });
+};
+
+export const followAction = username => (dispatch) => {
+  dispatch(pageLoadingAction());
+  const user = getCurrentUser();
+  return api
+    .post(`profiles/${username}/follow/`)
+    .then((res) => {
+      dispatch(followSuccess(res.data));
```
Comment (paulrimiru) [inline src/containers/FollowUnfollow/state/actions.js:65]: Add tests for the dispatches, ask @mutaimwiti to help

File: src/containers/FollowUnfollow/state/actions.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      dispatch(pageLoadedAction());
+    });
+};
+
+export const unfollowAction = username => (dispatch) => {
+  const user = getCurrentUser();
+  return api.delete(`profiles/${username}/follow/`)
+    .then((res) => {
+      dispatch(UnfollowSuccess(res.data));
+      dispatch(pageLoadedAction());
```
Comment (paulrimiru) [inline src/containers/FollowUnfollow/state/actions.js:81]: Same case here

**Review by paulrimiru — CHANGES_REQUESTED** [review]
Please make these changes before EOD so that we can merge this

File: src/containers/FollowUnfollow/state/actions.js (hunk 6 above)
Comment (C3real-kill3r) [reply src/containers/FollowUnfollow/state/actions.js:50]: Okay, on it

File: src/containers/FollowUnfollow/index.js (hunk 5 above)
Comment (C3real-kill3r) [reply src/containers/FollowUnfollow/index.js]: Okay

File: src/containers/FollowUnfollow/index.js (hunk 4 above)
Comment (C3real-kill3r) [reply src/containers/FollowUnfollow/index.js:19]: Okay

File: src/containers/FollowButton/index.js (hunk 3 above)
Comment (C3real-kill3r) [reply src/containers/FollowButton/index.js:18]: Noted

File: src/containers/ArticleListing/state/reducer.js (hunk 2 above)
Comment (C3real-kill3r) [reply src/containers/ArticleListing/state/reducer.js:48]: Seen

File: src/components/ProfileView/index.js (hunk 1 above)
Comment (C3real-kill3r) [reply src/components/ProfileView/index.js:8]: Cool....noted

**Review by bevkololi — APPROVED** [review]

**Review by paulrimiru — APPROVED** [review]
